tracer/src/Datadog.Tracer.Native/debugger_method_rewriter.cpp

```diff
@@ -758,6 +714,456 @@ HRESULT DebuggerMethodRewriter::ApplyMethodProbe(
+
+    EHClause beginMethodExClause{};
+    beginMethodExClause.m_Flags = COR_ILEXCEPTION_CLAUSE_NONE;
+    beginMethodExClause.m_pTryBegin = firstInstruction;
+    beginMethodExClause.m_pTryEnd = beginMethodCatchFirstInstr;
+    beginMethodExClause.m_pHandlerBegin = beginMethodCatchFirstInstr;
+    beginMethodExClause.m_pHandlerEnd = beginMethodCatchLeaveInstr;
+    beginMethodExClause.m_ClassToken = debuggerTokens->GetExceptionTypeRef();
+    newClauses.push_back(beginMethodExClause);
+
+    ILInstr* beginOriginalMethodInstr = rewriterWrapper.GetCurrentILInstr();
+    beginMethodTryLeaveInstr->m_pTarget = beginOriginalMethodInstr;
+    beginMethodCatchLeaveInstr->m_pTarget = beginOriginalMethodInstr;
+
+    // ***
+    // ENDING OF THE METHOD EXECUTION
+    // ***
+
+    ILInstr* endMethodOriginalCodeFirstInstr = nullptr;
+    ILInstr* endMethodTryStartInstr = nullptr;
+    hr = EndAsyncMethodProbe(corProfiler, rewriterWrapper, moduleMetadata, debuggerTokens, caller, isStatic,
+                             methodReturnType, methodLocals, numLocals, asyncMethodStateIndex, callTargetReturnIndex,
+                             newClauses, &endMethodTryStartInstr, &endMethodOriginalCodeFirstInstr);
+
+    if (FAILED(hr))
+    {
+        Logger::Error("DebuggerMethodRewriter::ApplyAsyncMethodProbe: Fail in EndAsyncMethodProbe");
+        return hr;
+    }
+
+    if (endMethodOriginalCodeFirstInstr == nullptr && endMethodTryStartInstr == nullptr)
+    {
+        // in this case there is no SetResult in the MoveNext method so we don't have to fix branches
+        return S_OK;
+    }
+
+    // Changes all LEAVE's to the original end method to the try end method
+    for (ILInstr* pInstr = rewriterWrapper.GetILRewriter()->GetILList()->m_pNext; 
+        pInstr != endMethodTryStartInstr; 
+        pInstr = pInstr->m_pNext)
+    {
+        switch (pInstr->m_opcode)
+        {
+            case CEE_LEAVE:
+            case CEE_LEAVE_S:
+            {
+                if (pInstr->m_pTarget == endMethodOriginalCodeFirstInstr)
+                {
+                    pInstr->m_pTarget = endMethodTryStartInstr;
+                }
+                break;
+            }
+            default:
+                break;
+        }
+    }
+
+    return S_OK;
+}
+
+HRESULT DebuggerMethodRewriter::IsTypeImplementIAsyncStateMachine(const ComPtr<IMetaDataImport2>& metadataImport,
+                                                                  const ULONG32 typeToken, bool& isTypeImplementIAsyncStateMachine)
+{
+    HCORENUM interfaceImplsEnum = nullptr;
+    ULONG actualImpls;
+    mdInterfaceImpl impls;
+    // check if the nested type implement the IAsyncStateMachine interface
+    const auto hr = metadataImport->EnumInterfaceImpls(&interfaceImplsEnum, typeToken, &impls, 1, &actualImpls);
+    if (hr != S_OK)
+    {
+        // FAILED or S_FALSE
+        return hr;
+    }
+
+    metadataImport->CloseEnum(interfaceImplsEnum);
```

Comment:
I think you have to call `CloseEnum` also if it returns `S_FALSE`. But I might be wrong. I suggest you to double check it and make sure we don't cause the enumeration to hang in memory.

Reply:
Fixed in 82878f89777debace4f5f6ff669306f87e4c1ae7